### PR TITLE
Writing/frontend

### PR DESCRIPTION
**Pull request by nico1207**
Folgende Kapitel hinzugefügt:

Grundlagen/Angular
Design/Frontend
Implementierung/WebAssembly-Interface
Implementierung/Frontend

### CODE REVIEW

File: latex/Inhalt/implementierung.tex [hunk 1]
```diff
@@ -543,3 +543,152 @@ \subsubsection{Rangierbahnhof-Algorithmus}
...
+
+\begin{minted}{typescript}
+    this._loop = window.setInterval(() => {
+      for (let i = 0; i < this._stepsPerInterval; i++) {
+        this.cpuStep();
+      }
+    }, this._interval);
+\end{minted}
+
+Statt einer \rust{while}-Schleife kommt hier die Funktion \ts{window.setInterval()} des Browsers zum Einsatz. Diese Funktion führt eine gegebene andere Funktion periodisch immer wieder mit einem bestimmten Intervall auf. Dies ermöglicht der Webanwendung trotz der \enquote{Endlossschleife} responsiv zu bleiben. Da der Intervall nicht beliebig klein sein kann, gibt es außerdem die Möglichkeit, in jedem \enquote{Schleifendurchlauf} mehrere Prozessorschritte durchzuführen. So kann der Emulator trotzdem so schnell laufen wie es nötig ist.
```
Comment (moorts) [inline latex/Inhalt/implementierung.tex:632]: "Das Intervall"

**Review by moorts — CHANGES_REQUESTED** [review]
Du hast irgendwo "assembliert" verwendet und woanders "assembled", das sollte man vermutlich vereinheitlichen

File: latex/Inhalt/implementierung.tex (hunk 1 above)
Comment (felixhir) [inline latex/Inhalt/implementierung.tex:632]: "das Intervall"

unquote "Schleifendurchlauf"

File: latex/Inhalt/implementierung.tex [hunk 2]
```diff
@@ -543,3 +543,152 @@ \subsubsection{Rangierbahnhof-Algorithmus}
...
+
+Die \ts{cpuStep()}-Methode arbeitet nun verschiedene Schritte ab. Hauptaufgabe der Methode ist das Ausführen von \ts{emulator.execute_next()}, was dafür sorgt, dass der Emulator die nächste Instruktion ausführt und den Program Counter gegebenenfalls erhöht. Nach dem Ausführen des Emulatorschritts wird der Arbeitsspeicher mit dem vorherigen Zustand verglichen, um festzustellen ob es eine Änderung gab. Dies wird für das Einfärben der entsprechenden Stelle in der RAM-Matrix genutzt.
```
Comment (felixhir) [inline latex/Inhalt/implementierung.tex:634]: end the first sentence with a colon.

RAM should be an acronym

File: latex/Inhalt/implementierung.tex [hunk 3]
```diff
@@ -543,3 +543,152 @@ \subsubsection{Rangierbahnhof-Algorithmus}
...
+
+Eine weitere Aufgabe der Methode ist die Emulation von sogenannten \textit{CP/M Syscalls}.
+
+\subsection{CP/M Emulation}
+
+Um den Emulator zu testen, wurden einige frei verfügbare Programme genutzt. Diese wurden allerdings für das CP/M Betriebssystem geschrieben, welches ab 1974 unter anderem für den Intel 8080 entwickelt wurde und auf diversen Rechnern für Endanwender zum Einsatz kam.
```
Comment (felixhir) [inline latex/Inhalt/implementierung.tex:640]: In this and the previous block CP/M should be an acronym

File: latex/Inhalt/implementierung.tex [hunk 4]
```diff
@@ -543,3 +543,152 @@ \subsubsection{Rangierbahnhof-Algorithmus}
...
+
+Ein wesentliches Feature, was diese Programme nutzen, sind die BDOS Syscalls, die auf einem CP/M System das Ausgeben von Text in einem Terminal ermöglichen \cite{cpmSyscalls}.
```
Comment (felixhir) [inline latex/Inhalt/implementierung.tex:642]: BDOS and CPM should be acronyms

"CP/M-System"

File: latex/Inhalt/implementierung.tex [hunk 5]
```diff
@@ -543,3 +543,152 @@ \subsubsection{Rangierbahnhof-Algorithmus}
...
+
+\begin{minted}{asm}
+    LD DE   ;parameter
+    LD C    ;function
+    CALL 5  ;syscall routine
+\end{minted}
+
+Für den Aufruf eines Syscalls muss die Funktion, die aufgerufen wird, in das Register C geschrieben werden. Mögliche Parameter können im Register DE abgelegt werden. Der Syscall kann nun durch einen Sprung auf Adresse 5 ausgeführt werden.
+
+Relevant für die Programme, die zum Testen genutzt werden, sind lediglich die Syscalls 2 (Zeichen-Ausgabe) und 9 (String-Ausgabe). Die Emulation dieser Syscalls ist vergleichsweise simpel.
```
Comment (felixhir) [inline latex/Inhalt/implementierung.tex:652]: End this block with a colon to refer to the following code

File: latex/Inhalt/implementierung.tex [hunk 6]
```diff
@@ -543,3 +543,152 @@ \subsubsection{Rangierbahnhof-Algorithmus}
...
+
+\begin{minted}{typescript}
+    if (this._cpmMode) {
+      // Patch memory for BDOS syscalls
+      this._emulatorMemory[5] = 0xC9; // RET
+      this._emulator.pc = 0x100;
+    }
+\end{minted}
+
+Beim Start eines Programms, was im CP/M Modus ausgeführt wird, wird der Speicher an der Stelle 5 mit einer Return-Instruktion überschrieben. Das sorgt dafür, dass das Programm bei jedem Syscall sofort wieder an die ursprüngliche Stelle zurückspringt. Zusätzlich wird der Program Counter beim Start direkt auf 0x100 gesetzt, da CP/M davon ausgeht, dass Nutzerprogramme an dieser Stelle beginnen.
```
Comment (felixhir) [inline latex/Inhalt/implementierung.tex:662]: CP/M should be an acronym

File: latex/Inhalt/implementierung.tex [hunk 7]
```diff
@@ -543,3 +543,152 @@ \subsubsection{Rangierbahnhof-Algorithmus}
...
+                    currentChar = String.fromCharCode(charCode);
+                }
+                console.log(fullString);
+                break;
+            }
+        }
+    }
+\end{minted}
+
+Sobald der Program Counter den Wert 5 hat, das Programm also an die Syscall Adresse gesprungen ist, wird dieser emuliert. Beim Syscall 2 wird das Zeichen, was sich Register E befindet, ausgegeben. Syscall 9 gibt den String, der sich an der Speicheradresse aus Register DE befindet, so lange aus, bis das Zeichen \enquote{\$} auftritt.
```
Comment (felixhir) [inline latex/Inhalt/implementierung.tex:694]: "wird dieser emuliert", what does "dieser" refer to? The code, or the syscall? Should be clarified

"was sich in Register E"

**Review by felixhir — CHANGES_REQUESTED** [review]
LineHighlight.png and Completion2.png are hard to read in a printed format. Have you tried cutting off some of the large blank spaces to the right of the images? IMO they don't add a lot of information

**Review by felixhir — APPROVED** [review]

**Pull request merged**